### PR TITLE
[PPR-356] Infra for Advanced Fees Management PoC

### PR DESCRIPTION
**Pull request by aomegax**
This PR is for creating a base infrastructure for the PoC regarding marketplace for advanced fees management (for more info 👉 [here](https://pagopa.atlassian.net/wiki/spaces/PPR/pages/511443239/GEC+-+Design+Review+-+Marketplace)). 

At the moment, the needs are:
- an app service to deploy java+spring applicative
- a cosmosdb to store data

The infra should be created only in dev env and, in the next future, it will be revised to adopt AKS.

### Type of changes

- [x] Add new resources
- [ ] Update configuration to existing resources
- [ ] Remove existing resources

### Does this introduce a change to production resources with possible user impact?

- [ ] Yes, users may be impacted applying this change
- [ ] No

### Does this introduce an unwanted change on infrastructure? Check terraform plan execution result

- [ ] Yes
- [ ] No

### Other information

---

### CODE REVIEW

File: src/advanced_fees_management.tf [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+  advanced_fees_management_cors_configuration = {
+    origins = ["*"]
+    methods = ["*"]
+  },
+  advanced_fees_management_cosmosdb_enable_serverless = contains(var.advanced_fees_management_cosmosdb_extra_capabilities, "EnableServerless")
+}
+
+# subnet
+module "advanced_fees_management_snet" {
+  count                                          = var.advanced_fees_management_enabled && var.cidr_subnet_advanced_fees_management != null ? 1 : 0
```
Comment (pasqualespica) [inline src/advanced_fees_management.tf:20]: I'd be more explicit I'd test env variable directly. It's a PoC

File: src/advanced_fees_management.tf [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+      name    = "Microsoft.Web/serverFarms"
+      actions = ["Microsoft.Network/virtualNetworks/subnets/action"]
+    }
+  }
+}
+
+# afm app configuration
+
+# app service: this is used only for PoC
+module "advanced_fees_management_app_service" {
```
Comment (pasqualespica) [inline src/advanced_fees_management.tf:42]: ok , but DR mention AKS

File: src/advanced_fees_management.tf [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    TIMEOUT_DELAY                                   = 300
+    # Integration with private DNS (see more: https://docs.microsoft.com/en-us/answers/questions/85359/azure-app-service-unable-to-resolve-hostname-of-vi.html)
+    WEBSITE_ADD_SITENAME_BINDINGS_IN_APPHOST_CONFIG = "1"
+    WEBSITE_RUN_FROM_PACKAGE                        = "1"
+    WEBSITE_VNET_ROUTE_ALL                          = "1"
+    WEBSITE_DNS_SERVER                              = "168.63.129.16"
+
+    # Spring Environment
+    COSMOS_KEY1 = data.azurerm_key_vault_secret.afm_cosmos_key1.value
+    COSMOS_KEY2 = data.azurerm_key_vault_secret.afm_cosmos_key2.value
```
Comment (pasqualespica) [inline src/advanced_fees_management.tf:90]: key2 too ?

File: src/advanced_fees_management.tf [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+  plan_name     = format("%s-plan-afm", local.project)
+  plan_kind     = "Linux"
+  plan_sku_tier = var.advanced_fees_management_tier
+  plan_sku_size = var.advanced_fees_management_size
+  plan_reserved = true # Mandatory for Linux plan
+
+  # App service plan
+  name                = format("%s-app-afm", local.project)
+  client_cert_enabled = false
+  always_on           = false
```
Comment (pasqualespica) [inline src/advanced_fees_management.tf:60]: evaluate if it's better set this one to `true` in dev too

File: src/advanced_fees_management.tf [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+# afm cosmos configuration
+
+# cosmosdb account
+module "advanced_fees_management_cosmosdb_account" {
+  count  = var.advanced_fees_management_enabled ? 1 : 0
+
+  source   = "git::https://github.com/pagopa/azurerm.git//cosmosdb_account?ref=v2.1.18"
+  name     = format("%s-afm-cosmos", local.project)
```
Comment (pasqualespica) [inline src/advanced_fees_management.tf:118]: rename with `-<YOUR_NAME>-cosmos-account`

File: src/advanced_fees_management.tf [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+#      location          = "northeurope"
+#      failover_priority = 1
+#      zone_redundant    = true
+#    }
+#  ]
+
+  # for the PoC we are not interested to backup
+  backup_continuous_enabled = false
+
+  is_virtual_network_filter_enabled = false
```
Comment (pasqualespica) [inline src/advanced_fees_management.tf:154]: it is preferable set to `true` and then add to `allowed_virtual_network_subnet_ids` whoever the access permissions

File: src/api/afm_api/marketplace/v1/_base_policy.xml.tpl [hunk 7]
```diff
@@ -0,0 +1,100 @@
+<!--
+    IMPORTANT:
+    - Policy elements can appear only within the <inbound>, <outbound>, <backend> section elements.
```
Comment (pasqualespica) [inline src/api/afm_api/marketplace/v1/_base_policy.xml.tpl:3]: you need this policy for the this PoC ?

File: src/advanced_fees_management.tf [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+# cosmosdb database
+module "advanced_fees_management_cosmosdb_database" {
+  count  = var.advanced_fees_management_enabled ? 1 : 0
+
+  source              = "git::https://github.com/pagopa/azurerm.git//cosmosdb_sql_database?ref=v2.1.15"
+  name                = "db"
+  resource_group_name = azurerm_resource_group.advanced_fees_management_rg.name
+  account_name        = module.advanced_fees_management_cosmosdb_account[0].name
+}
```
Comment (pasqualespica) [inline src/advanced_fees_management.tf:178]: no `cosmosdb_sql_container` ? db only ?

**Review by pasqualespica — CHANGES_REQUESTED** [review]

File: src/advanced_fees_management.tf (hunk 2 above)
Comment (aomegax) [reply src/advanced_fees_management.tf:42]: yes, but there is a note that specifies we will use app service for time constraints (we have less than 10 days).
AKS will be the solution for the next future.

File: src/api/afm_api/marketplace/v1/_base_policy.xml.tpl (hunk 7 above)
Comment (aomegax) [reply src/api/afm_api/marketplace/v1/_base_policy.xml.tpl:3]: no, it is a wrong cut&paste 👍

File: src/advanced_fees_management.tf (hunk 8 above)
Comment (aomegax) [reply src/advanced_fees_management.tf:178]: yes, wip 🚧

File: src/advanced_fees_management.tf (hunk 1 above)
Comment (aomegax) [reply src/advanced_fees_management.tf]: done

File: src/advanced_fees_management.tf (hunk 3 above)
Comment (aomegax) [reply src/advanced_fees_management.tf]: done

**Review by pasqualespica — APPROVED** [review]
great 👍

**Pull request merged**